apollo-router/src/services/layers/persisted_queries/mod.rs

```diff
@@ -110,9 +115,32 @@ impl PersistedQueryLayer {
         } else {
             // if there is no query, look up the persisted query in the manifest
             // and put the body on the `supergraph_request`
-            if let Some(persisted_query_body) =
-                manifest_poller.get_operation_body(persisted_query_id)
-            {
+            if let Some(persisted_query_body) = manifest_poller.get_operation_body(
+                persisted_query_id,
+                // Use the first one of these that exists:
+                // - The PQL-specific context name entry
+                //   `apollo_persisted_queries::client_name` (which can be set
+                //   by router_service plugins)
+                // - The same name used by telemetry if telemetry is enabled
+                //   (ie, the value of the header named by
+                //   `telemetry.apollo.client_name_header`, which defaults to
+                //   `apollographql-client-name` by default)
+                // - The value in the `apollographql-client-name` header
+                //   (whether or not telemetry is enabled)
+                request
+                    .context
+                    .get(PERSISTED_QUERIES_CLIENT_NAME_CONTEXT_KEY)
+                    .unwrap_or_default()
+                    .or_else(|| request.context.get(CLIENT_NAME).unwrap_or_default())
+                    .or_else(|| {
+                        request
+                            .supergraph_request
+                            .headers()
+                            .get(client_name_header_default_str())
```

Comment:
This won't be right if the user has configured a custom client header name.
Telemetry puts the client name into the context. So why do we need to check for a header at all here?

Reply:
It seemed reasonable to have the feature not be completely broken if telemetry isn't enabled (which, eg, was the case in the integration tests). Eg if you're running your router locally with your client that's built to sent PQs. That said I feel like I'd be happier if I could get the implementation to be "honor `telemetry.apollo.client_name_header even if APOLLO_KEY isn't set" but I couldn't quite figure out a comfortable way to get access to the plugin configuration from outside of it.